src/librustc_lint/builtin.rs

```diff
@@ -1225,3 +1226,40 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnionsWithDropFields {
         }
     }
 }
+
+/// Lint for casting from non-{i,u}size integers to pointers.
+pub struct IntToPtrCast;
+
+declare_lint! {
+    INT_TO_PTR_CAST,
+    Warn,
+    "casting from non-{i,u}size integers to pointers"
+}
+
+impl LintPass for IntToPtrCast {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INT_TO_PTR_CAST)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for IntToPtrCast {
+    fn check_expr(&mut self, ctx: &LateContext, expr: &hir::Expr) {
+        if let hir::ExprCast(ref source, _) = expr.node {
+            if let Some(&CastKind::AddrPtrCast) = ctx.tables.cast_kinds.get(&source.id) {
+                let source_ty = &ctx.tables.expr_ty(&source).sty;
+                match source_ty {
+                    &ty::TyInt(ast::IntTy::Is) |
+                    &ty::TyUint(ast::UintTy::Us) => {},
+                    &ty::TyInt(_) |
+                    &ty::TyUint(_) => {
+                        ctx.span_lint(
+                            INT_TO_PTR_CAST,
+                            expr.span,
+                            "casting from a different-size integer to a pointer");
```

Comment:
The message is a placeholder, but I think "casting from a potentially non-pointer-sized integer to a pointer" would be better English.